src/pool.ts

```diff
@@ -0,0 +1,789 @@
+  }
+
+  public get connection(): AwaitConnection {
+    if (this._connection === null) {
+      throw new errors.InterfaceError("The proxy is detached");
+    }
+    return this._connection;
+  }
+
+  /**
+   * Returns a value indicating whether this PoolConnectionProxy is detached
+   * from a connection.
+   */
+  public get detached(): boolean {
+    return this._connection === null;
+  }
+
+  /**
+   * Returns a value indicating whether this PoolConnectionProxy is detached
+   * or the underlying connection is closed.
+   */
+  isClosed(): boolean {
+    return this._connection === null || this._connection.isClosed();
+  }
+
+  /** @internal */
+  onConnectionClose(): void {
+    this._holder._releaseOnClose();
+  }
+
+  /** @internal */
+  public get holder(): PoolConnectionHolder {
+    return this._holder;
+  }
+
+  /** @internal */
+  detach(): AwaitConnection | null {
+    if (this._connection === null) {
+      return null;
+    }
+
+    const connection = this._connection;
+    this._connection = null;
+    connection._setProxy(null);
+    return connection;
+  }
+}
+
+const DefaultMinPoolSize = 0;
+const DefaultMaxPoolSize = 100;
+
+export interface PoolOptions {
+  connectOptions?: ConnectConfig | null;
+  minSize?: number;
+  maxSize?: number;
+  onAcquire?: (proxy: PoolConnectionProxy) => Promise<void>;
+  onRelease?: (proxy: PoolConnectionProxy) => Promise<void>;
+  onConnect?: (connection: AwaitConnection) => Promise<void>;
+  connectionFactory?: (
+    options?: ConnectConfig | null
+  ) => Promise<AwaitConnection>;
+}
+
+/**
+ * A connection pool.
+ *
+ * Connection pool can be used to manage a set of connections to the database.
+ * Connections are first acquired from the pool, then used, and then released
+ * back to the pool. Once a connection is released, it's reset to close all
+ * open cursors and other resources *except* prepared statements.
+ * Pools are created by calling :func:`~edgedb.pool.create`.
+ */
+export class Pool implements IConnection {
+  private _closed: boolean;
+  private _closing: boolean;
+  private _queue: LifoQueue<PoolConnectionHolder>;
+  private _holders: PoolConnectionHolder[];
+  private _initialized: boolean;
+  private _initializing: boolean;
+  private _minSize: number;
+  private _maxSize: number;
+  private _onAcquire?: (proxy: PoolConnectionProxy) => Promise<void>;
+  private _onRelease?: (proxy: PoolConnectionProxy) => Promise<void>;
+  private _onConnect?: (connection: AwaitConnection) => Promise<void>;
+  private _connectionFactory: (
+    options?: ConnectConfig | null
+  ) => Promise<AwaitConnection>;
+  private _generation: number;
+  private _connectOptions?: ConnectConfig | null;
+
+  public get initialized(): boolean {
+    return this._initialized;
+  }
+
+  /** @internal */
+  public get generation(): number {
+    return this._generation;
+  }
+
+  /** @internal */
+  public get holders(): PoolConnectionHolder[] {
+    return this._holders;
+  }
```

Comment:
Are these getters used anywhere? Why are they public?

Reply:
You are right, we don't need these. I have the tendency of leaving things public and open, because I think: "You never know what users will need to do" (I like to think that Python doesn't have private modifier for a similar reason). I will remove these public getters :)

Reply:
On a second thought, having the queue public might have a good use case: consider if somebody implements telemetry and wants to monitor the size of the queue. And `generation` is used by the ConnectionHolder, so it makes sense to leave it as public get only property (public is actually redundant since things are public by default anyway in TS)

Reply:
The only way I was using `holders` was to verify that a pool is created with a `minSize` of open connections. This seems a valid use case to me, too.

```ts
describe("pool.initialize: creates minSize count of connections", () => {
  each([0, 1, 5, 10, 20]).it(
    "when minSize is '%s'",
    async (minSize) => {
      const maxSize = minSize + 50;
      const pool = await Pool.create({
        connectOptions: getConnectOptions(),
        minSize,
        maxSize,
      });

      expect(pool.queue).toHaveLength(maxSize);

      const itemsWithOpenConnection = pool.holders.filter(
        (holder) => holder.connection !== null
      );

      expect(itemsWithOpenConnection).toHaveLength(minSize);
```

Reply:
Now I recall one of the reasons why I left `holders` and `holder` public: the `ConnectionProxy` is exposed by `onAcquire` and `onRelease` callbacks, and the proxy's holder is read inside a pool, to ensure that the proxy belongs to the pool where it's being released.
1. I used public getter because I wanted certain properties to be read-only, and didn't take into consideration something like:

```ts
private __something: T;  // *really* private

/** @internal */
get _something(): T {
  return this.__something;
}
```

2. An alternative would be to replace the get-only property with a method like:

```
/** @internal */
_getHolder() {
  return this._holder;
}
```